lua/neogit/status.lua

```diff
@@ -64,390 +69,191 @@ local function get_section_item_for_line(linenr)
-      i.diff_open = false
-    end
-  end
+  status_buffer:unlock()
 
-  local line = vim.fn.line(".")
-  local section_idx = get_section_idx_for_line(line)
-  local section = locations[section_idx]
+  draw_buffer()
+  draw_signs()
 
-  display_status()
+  status_buffer:lock()
+end
 
+local function toggle()
+  local section, item = get_current_section_item()
   if section == nil then
-    primitive_move_cursor(line)
     return
   end
 
-  local item_idx = line - section.first
-
-  if section.name == "unstaged_changes" then
-    if contextually_move_cursor("unstaged_changes", "staged_changes", item_idx) then
-      return
-    end
-  elseif section.name == "staged_changes" then
-    if contextually_move_cursor("staged_changes", "unstaged_changes", item_idx) then
-      return
-    end
-  elseif section.name == "untracked_files" then
-    if contextually_move_cursor("untracked_files", "staged_changes", item_idx) or
-       contextually_move_cursor("untracked_files", "unstaged_changes", item_idx) then
-     return
-   end
-  end
-  primitive_move_cursor(line)
-end
-
-local load_diffs = a.sync(function ()
-  local unstaged = {}
-  local staged = {}
-  for _,c in pairs(status.unstaged_changes) do
-    if c.type ~= "Deleted" and c.type ~= "New file" and not c.diff_open and not c.diff_content then
-      table.insert(unstaged, c)
-    end
-  end
-  local unstaged_len = #unstaged
-  for _,c in pairs(status.staged_changes) do
-    if c.type ~= "Deleted" and c.type ~= "New file" and not c.diff_open and not c.diff_content then
-      table.insert(staged, c)
-    end
-  end
+  local on_hunk = item ~= nil and line_is_hunk(vim.fn.getline('.'))
 
-  local cmds = {}
-  for _, c in pairs(unstaged) do
-    if c.original_name ~= nil then
-      table.insert(cmds, cli.diff.files(c.original_name, c.name))
-    else
-      table.insert(cmds, cli.diff.files(c.name))
-    end
-  end
-  for _, c in pairs(staged) do
-    if c.original_name ~= nil then
-      table.insert(cmds, cli.diff.cached.files(c.original_name, c.name))
-    else
-      table.insert(cmds, cli.diff.cached.files(c.name))
-    end
-  end
-  local results = { a.wait(git.cli.in_parallel(unpack(cmds)).call()) }
+  if on_hunk then
+    local hunk = get_current_hunk_of_item(item)
+    hunk.folded = not hunk.folded
+  elseif item then item.folded = not item.folded
+  else section.folded = not section.folded end
 
-  for i=1,unstaged_len do
-    local name = unstaged[i].name
-    for _,c in pairs(status.unstaged_changes) do
-      if c.name == name then
-        c.diff_content = git.diff.parse(vim.split(results[i], '\n'))
-        break
-      end
-    end
-  end
-  for i=unstaged_len+1,#results do
-    local name = staged[i - unstaged_len].name
-    for _,c in pairs(status.staged_changes) do
-      if c.name == name then
-        c.diff_content = git.diff.parse(vim.split(results[i], '\n'))
-        break
-      end
-    end
-  end
-end)
+  refresh_status()
+end
 
+function __NeogitStatusReset()
```

Comment:
Also do you think we should remove these global `__...` functions and refactor the code that uses these to just use something like this: `require('neogit').status.reset()` ?

Reply:
Totally agree. We do need true globals sometimes (`v:lua` only takes globals), but I much prefer the `__NEOGIT` object I started for that anyway, since it scales better in that regard.